internal/application/application.go

```diff
@@ -494,8 +568,127 @@ func Logs(ctx context.Context, logChan chan tailer.ContainerLogLine, wg *sync.Wa
-// application from the cluster.
+// aggregate is an internal helper for List. It merges the information from an application resource
+// and adjacent secrets, pods, metrics, etc. into a proper application structure.
+
+func aggregate(ctx context.Context,
+	cluster *kubernetes.Cluster,
+	appCR unstructured.Unstructured,
+	scales map[string]v1.Secret,
+	bounds map[string]v1.Secret,
+	envs map[string]v1.Secret,
+	pods map[string][]v1.Pod,
+	routes map[string][]string,
+	metrics map[string]metricsv1beta1.PodMetrics,
+) (*models.App, error) {
+	key := ConfigurationKey(appCR.GetName(), appCR.GetNamespace())
+
+	appPods := pods[key]
+	appRoutes := routes[key]
+
+	appName := appCR.GetName()
+	namespace := appCR.GetNamespace()
+	meta := models.NewAppRef(appName, namespace)
+	app := meta.App()
+
+	// I. Unpack the auxiliary data in the various secrets
+
+	env, found := envs[key]
+	if !found {
+		return nil, errors.New("finding env")
+	}
+	environment := EnvironmentFromSecret(&env)
+
+	scale, found := scales[key]
+	if !found {
+		return nil, errors.New("finding scaling")
+	}
+	instances, err := ScalingFromSecret(&scale)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding scaling")
+	}
+
+	config, found := bounds[key]
+	if !found {
+		return nil, errors.New("finding configurations")
+	}
+	configurations := BoundConfigurationNamesFromSecret(&config)
+
+	// II. Unpack the core application resource
+
+	origin, err := Origin(&appCR)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding origin")
+	}
+
+	chartName, err := AppChart(&appCR)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding app chart")
+	}
+
+	stageID, err := StageID(&appCR)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding the stage id")
+	}
+
+	imageURL, err := ImageURL(&appCR)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding the image url")
+	}
+
+	settings, err := Settings(&appCR)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding settings")
+	}
+
+	desiredRoutes, err := DesiredRoutes(&appCR)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding desired routes")
+	}
+
+	// III. Assemble the main structure
+
+	app.Meta.CreatedAt = appCR.GetCreationTimestamp()
+
+	app.Configuration.Instances = &instances
+	app.Configuration.Configurations = configurations
+	app.Configuration.Environment = environment
+	app.Configuration.Routes = desiredRoutes
+	app.Configuration.AppChart = chartName
+	app.Configuration.Settings = settings
+	app.Origin = origin
+	app.StageID = stageID
+	app.ImageURL = imageURL
+
+	// IV. Assemble the deployment structure for active applications.
+
+	var status string
+	podMetrics := []metricsv1beta1.PodMetrics{}
+
+	if metrics == nil {
+		status = "failed to get replica details"
+	} else {
+		// extract the metrics for the app, based on the app pods
+		for _, pod := range appPods {
+			m, found := metrics[pod.Name]
+			if found {
+				podMetrics = append(podMetrics, m)
+			}
+		}
+	}
```

Comment:
I guess we can drop the else check, since it's safe to access a `nil` map (you cannot write but you can check the value on it)

_untested code!_

```go
var status string
if metrics == nil {
	status = "failed to get replica details"
}

// extract the metrics for the app, based on the app pods
podMetrics := []metricsv1beta1.PodMetrics{}
for _, pod := range appPods {
	if m, found := metrics[pod.Name]; found {
		podMetrics = append(podMetrics, m)
	}
}
```

Reply:
The new code wastes cycles on finding metrics for pod, where we know we have none.
The old code does not waste cycles as skips the loops entirely.
Would prefer to keep that.